### ISSUE TITLE
YOLOv8n Training (1차)

### ISSUE THREAD

**Issue by juooo1117** [jsj5605/ROADs#9]
### Dataset Pre-processing
1. YOLOv8 모델은 바운딩 박스를 (x_center, y_center, width, height) 형식으로 입력해야 하는데, 우리가 가지고 있는 annotation file 의 bbox format은 (좌상단 x좌표, 좌상단 y좌표, bbox_width, bbox_height)이므로 center point를 잡으면서 normalizing 하는 과정을 먼저 반영했다.

```python
    def convert_to_yolov8_format(bbox, image_width, image_height):

    x_center = (bbox[0] + bbox[2] / 2) / image_width
    y_center = (bbox[1] + bbox[3] / 2) / image_height
    width = bbox[2] / image_width
    height = bbox[3] / image_height

    return x_center, y_center, width, height
```

2. AI HUB dataset 에서 다운로드 받은 자료를 살펴보니, 이미지 한 장 당 annotation 정보를 담은 json 파일이 1개씩 존재한다. 따라서 각 json 파일에서 필요한 정보만 추출해서 사용하기로 결정하고 최종으로 필요한 json 파일 구조는 아래와 같다.
```
.json dataset
├─ image                  
│   ├─ id                 
│   ├─ width              # 1920
│   ├─ height             # 1080
│   └─ file_name          # real_sunny_AM_20220614_112456_sedan_Dong-gu_00026.jpg
└─ annotations                
    ├─ id                 # detected image id
    ├─ image_id           # original image id
    ├─ category_id        # class name
    └─ bbox               # 좌상단x, 좌상단y, bbox의 W, bbox의 H
```
3. YOLO v8n 전이 학습 기반 이미지 객체 바운딩 박스 검출 과정을 수행한다.

- 속도 및 정확도 측면에서 높은 성능을 보여 다양한 객체 검출에 활용되는 모델 활용
- 모델 성능 개선을 위해 MS COCO dataset 에서 사전 학습된 모델을 통해 전이학습 진행

### Result category

1. **box_loss**: 바운딩 박스의 손실(loss)을 나타냄. 작을수록 바운딩 박스 예측이 정확하다.
2. **cls_loss**: 클래스 예측의 손실을 나타냄. 작을수록 클래스 예측이 정확하다.
3. **dfl_loss**: YOLOv4의 대체 손실인 DFL(Dynamic Focal Loss)의 값
4. **Instances**: 현재 에폭에서 감지된 객체의 인스턴스 수
5. **Class**: 클래스의 이름
6. **Images**: 테스트에 사용된 이미지 수
7. **Instances**: 감지된 객체의 인스턴스 수
8. **Box(P, R, mAP50, mAP50-95)**: Precision, Recall, mAP at 50, mAP at 50-95

    |Epoch|box_loss|cls_loss|dfl_loss|Instances|Class|Images|Instances|Box(P)|Box(R)|Box(mAP50,)|Box(mAP50-95)|
    |------|---------|--------|--------|----------|-----|-------|----------|-------|------|--------------|----------------|
    |1/20|0.8356|6.774|1.044|16|all|4000|7238|0.34|0.337|0.252|0.199|
    |2/20|0.8257|3.34|1.018|15|all|4000|7238|0.407|0.408|0.366|0.298|
    |3/20|0.8254|1.997|1.023|13|all|4000|7238|0.592|0.374|0.382|0.307|
    |4/20|0.7813|1.555|0.9955|13|all|4000|7238|0.541|0.508|0.501|0.411|
    | ... | ... | ... | ... | ... | ... | ... | ... | ... | ... | ... |... |
    |17/20|0.4911|0.4422|0.8533|15|all|4000|7238|0.903|0.861|0.894|0.827|
    |18/20|0.4785|0.4217|0.8486|16|all|4000|7238|0.916|0.881|0.926|0.862|
    |19/20|0.4662|0.401|0.8426|14|all|4000|7238|0.917|0.885|0.909|0.845|
    |20/20|0.4554|0.3828|0.8377|14|all|4000|7238|0.932|0.891|0.932|0.873|

### Image resizing 관련
YOLOv8 모델의 경우 정사각형 이미지만 Input으로 받는다는 것을 확인 → 우리가 가진 dataset의 경우는 (1920, 1080) 사이즈의 이미지이기 때문에 정사각형으로 resizing을 고려했다.
하지만, training 과정에서 'rect=True' 로 설정해 주면 [직사각형 형태도 넣을 수 있다는 것을 확인](https://github.com/ultralytics/ultralytics/issues/235)했고, 아래와 같은 조건으로 학습을 진행했다.
```python
    !yolo task=detect 
             mode=train 
             model=yolov8n.pt 
             data=/content/drive/MyDrive/FinalProject/data.yaml 
             epochs=20 
             imgsz=[1920,1080] 
             rect=True
```

### Resuming Training with the last.pt file
20000장의 이미지로 모델 훈련하는 과정에서 1epoch 당 거의 40분 정도가 소요되었다. 따라서 총 20epoch을 돌렸기 때문에 전체 epoch을 한 번에 훈련시킬 수 없어서 학습을 끊었다가 다시 재개하는 과정이 필요했다. 아래 코드를 이용하면 이전에 학습 완료했던 YOLOv8n 모델의 weight 들을 가지고 이어서 학습을 재개할 수 있었다.
```python
model = YOLO("/content/drive/MyDrive/FinalProject/runs/detect/train2/weights/last.pt")
model.train(resume=True)
```
